miniapp/miniapp_tsgemm.cpp

```diff
@@ -0,0 +1,473 @@
+  // 4. Wait for all
+  //
+  for (int i = 0; i < ps.num_iters; ++i) {
+    MPI_Barrier(MPI_COMM_WORLD);
+    auto t_start = clock_t::now();
+
+    schedule_gemm(a_mat, b_mat, cini_mat);
+    schedule_comm(mpi_executor, comm_grid, cfin_dist, cini_mat);
+    schedule_offload(cini_mat, cfin_mat);
+    waitall_tiles(cfin_mat);
+
+    MPI_Barrier(MPI_COMM_WORLD);
+    auto t_end = clock_t::now();
+
+    if (rank == 0) {
+      std::printf("%d: t_tot  [s] = %.5f\n", i, seconds_t(t_end - t_start).count());
+    }
+
+    //    // Simple check
+    //    ScalarType cfin_sum = sum_matrix(comm_world, cfin_mat);
+    //    if (rank == 0) {
+    //      std::cout << cfin_sum << '\n';
+    //    }
+  }
+
+  // Upon exit the mpi/user polling RAII object will stop polling
+  return hpx::finalize();
+}
+
+// Example usage:
+//
+//   mpirun -np 1 tsgemm --len_m      100  --len_n      100  --len_k  10000
+//                       --tile_m      64  --tile_n
+//                       --pgrid_rows   1  --pgrid_cols   1
+//
+int main(int argc, char** argv) {
+  // Flush printf
+  setbuf(stdout, nullptr);
+
+  // Initialize MPI
+  int thd_required = MPI_THREAD_MULTIPLE;
+  int thd_provided;
+  MPI_Init_thread(&argc, &argv, thd_required, &thd_provided);
+
+  if (thd_required != thd_provided) {
+    std::fprintf(stderr, "Provided MPI threading model does not match the required one.\n");
+    MPI_Abort(MPI_COMM_WORLD, 1);
+  }
+
+  // declare options before creating resource partitioner
+  ::options_description desc_cmdline = ::init_desc();
+
+#ifdef DLAF_WITH_MPI_POOL
+  // NB.
+  // thread pools must be declared before starting the runtime
+
+  // Create resource partitioner
+  hpx::resource::partitioner rp(desc_cmdline, argc, argv);
+
+  // create a thread pool that is not "default" that we will use for MPI work
+  rp.create_thread_pool("mpi");
+
+  // add (enabled) PUs on the first core to it
+  rp.add_resource(rp.numa_domains()[0].cores()[0].pus(), "mpi");
+#endif
+
+  // Start the HPX runtime
+  hpx::init(desc_cmdline, argc, argv);
+
+  MPI_Finalize();
+}
+
+namespace {
+
+void schedule_gemm(ConstMatrixType& a_mat, ConstMatrixType& b_mat, MatrixType& cini_mat) {
+  // TODO: Order communication and computation to limit the number of simultaneous non-blocking MPI calls
+
+  dlaf::LocalTileSize const& tile_size = cini_mat.distribution().localNrTiles();
+  for (SizeType tile_j = 0; tile_j < tile_size.cols(); ++tile_j) {
+    for (SizeType tile_i = 0; tile_i < tile_size.rows(); ++tile_i) {
+      LocalTileIndex a_mat_idx(0, tile_i);
+      LocalTileIndex b_mat_idx(0, tile_j);
+      LocalTileIndex c_mat_idx(tile_i, tile_j);
+
+      auto gemm_f = hpx::util::unwrapping([](auto&& a_tile, auto&& b_tile, auto&& c_tile) {
+        dlaf::tile::gemm(blas::Op::Trans, blas::Op::NoTrans, ScalarType(1), a_tile, b_tile,
+                         ScalarType(0), c_tile);
+      });
+      hpx::dataflow(gemm_f, a_mat.read(a_mat_idx), b_mat.read(b_mat_idx), cini_mat(c_mat_idx));
```

Comment:
You can also use directly the dataflow without the explicit lambda like this:

```cpp
hpx::dataflow(
  hpx::util::unwrapping(dlaf::tile::gemm),
  blas::Op::Trans, blas::Op::NoTrans, ScalarType(1),
  a_mat.read(a_mat_idx),
  b_mat.read(b_mat_idx),
  ScalarType(0), cini_mat(c_mat_idx));
```

Reply:
I prefer the lambdas as I don't like placing too many parameters in `hpx::dataflow`. I find it easier to read and reason about.